File: cpp/70. Climbing Stairs.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Down-Top DP like Fibonacci, Time: O(n), Space: O(1)
public:
    int climbStairs(int n) {
        if(n==1) return 1;
        if(n==2) return 2;
        int prevAns = 1;
        int curAns = 2;
        for(int i=3; i<=n; i++) {
            int tmpAns = curAns;
            curAns = prevAns + curAns;
            prevAns = tmpAns;
        }
        return curAns;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; int n; int ans;

    solution = Solution();
    n = 2;
    ans = solution.climbStairs(n);
    assert(ans==2);

    solution = Solution();
    n = 3;
    ans = solution.climbStairs(n);
    assert(ans==3);
}

File: python/70-Climbing Stairs.py
from functools import lru_cache

'''
dp[n]: number of dinstinct ways to climb n stairs
dp[n] = dp[n-1] + dp[n-2]
dp[1] = 1
dp[2] = 2
'''

class Solution:
    # @lru_cache(None)
    # def climbStairs(self, n: int) -> int:
    #     "Top-Down with Memoization, Time: O(n), Space: O(n)"
    #     if n == 1: return 1
    #     if n == 2: return 2
    #     return self.climbStairs(n-1) + self.climbStairs(n-2)
    # def climbStairs(self, n: int) -> int:
    #     "Down-Top dp, Time: O(n), Space: O(n) can be reduce to O(1)"
    #     if n == 1: return 1
    #     if n == 2: return 2
    #     dp = [0] * (n+1)
    #     dp[1] = 1
    #     dp[2] = 2
    #     for i in range(3, n+1):
    #         dp[i] = dp[i-1] + dp[i-2]
    #     return dp[n]
    def climbStairs(self, n: int) -> int:
        "Down-Top dp, Time: O(n), Space: O(1)"
        if n == 1: return 1
        if n == 2: return 2
        pre = 1
        cur = 2
        for i in range(3, n+1):
            cur, pre = cur+pre, cur
        return cur

n = 2
result = Solution().climbStairs(n)
assert result == 2

n = 3
result = Solution().climbStairs(n)
assert result == 3